// @tsparticles/preset-matrix/tsparticles.preset.matrix.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Preset v4.0.2 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/basic'), require('@tsparticles/shape-matrix'), require('@tsparticles/plugin-poisson-disc'), require('@tsparticles/effect-shadow'), require('@tsparticles/plugin-trail')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/basic', '@tsparticles/shape-matrix', '@tsparticles/plugin-poisson-disc', '@tsparticles/effect-shadow', '@tsparticles/plugin-trail'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.presets = global.__tsParticlesInternals.presets || {}, global.__tsParticlesInternals.presets.matrix = global.__tsParticlesInternals.presets.matrix || {}), global.__tsParticlesInternals.bundles.basic, global.__tsParticlesInternals.shapes.matrix, global.__tsParticlesInternals.plugins.poisson, global.__tsParticlesInternals.effects.shadow, global.__tsParticlesInternals.plugins.trail));
})(this, (function (exports, basic, shapeMatrix, pluginPoissonDisc, effectShadow, pluginTrail) { 'use strict';

    const options = {
        fpsLimit: 30,
        particles: {
            number: {
                value: 300,
                density: {
                    enable: true,
                },
            },
            paint: {
                fill: {
                    color: {
                        value: {
                            h: 120,
                            s: 100,
                            l: 50,
                        },
                        animation: {
                            l: {
                                enable: true,
                                speed: 70,
                                sync: false,
                                min: 30,
                                max: 100,
                            },
                        },
                    },
                    enable: true,
                },
            },
            shape: {
                type: "matrix",
                options: {
                    matrix: {
                        interval: {
                            min: 30,
                            max: 60,
                        },
                    },
                },
            },
            effect: {
                type: "shadow",
                options: {
                    shadow: {
                        color: "#00ff41",
                        blur: 7,
                    },
                },
            },
            size: {
                value: 10,
            },
            move: {
                enable: true,
                direction: "bottom",
                straight: true,
                speed: {
                    min: 8,
                    max: 15,
                },
            },
        },
        trail: {
            enable: true,
            length: 10,
            fill: {
                color: "#000",
            },
        },
        poisson: {
            enable: true,
        },
        background: {
            color: "#000000",
        },
    };

    const presetName = "matrix";
    async function loadMatrixPreset(engine) {
        await engine.pluginManager.register(async (e) => {
            await Promise.all([
                basic.loadBasic(e),
                effectShadow.loadShadowEffect(e),
                pluginTrail.loadTrailPlugin(e),
                pluginPoissonDisc.loadPoissonDiscPlugin(e),
                shapeMatrix.loadMatrixShape(e),
            ]);
            e.pluginManager.addPreset(presetName, options);
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadMatrixPreset = loadMatrixPreset;

    exports.loadMatrixPreset = loadMatrixPreset;

}));
Object.assign(globalThis.window || globalThis, { loadMatrixPreset: (globalThis.__tsParticlesInternals.presets.matrix || {}).loadMatrixPreset });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
