// @tsparticles/plugin-trail/tsparticles.plugin.trail.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Plugin v4.0.2 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.plugins = global.__tsParticlesInternals.plugins || {}, global.__tsParticlesInternals.plugins.trail = global.__tsParticlesInternals.plugins.trail || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    class TrailFill {
        color;
        image;
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.color !== undefined) {
                this.color = engine.OptionsColor.create(this.color, data.color);
            }
            if (data.image !== undefined) {
                this.image = data.image;
            }
        }
    }

    class Trail {
        enable;
        fill;
        length;
        constructor() {
            this.enable = false;
            this.length = 10;
            this.fill = new TrailFill();
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.fill !== undefined) {
                this.fill.load(data.fill);
            }
            if (data.length !== undefined) {
                this.length = data.length;
            }
        }
    }

    class TrailPlugin {
        id = "trail";
        _pluginManager;
        constructor(pluginManager) {
            this._pluginManager = pluginManager;
        }
        async getPlugin(container) {
            const { TrailPluginInstance } = await Promise.resolve().then(function () { return TrailPluginInstance$1; });
            return new TrailPluginInstance(this._pluginManager, container);
        }
        loadOptions(_container, options, source) {
            if (!this.needsPlugin()) {
                return;
            }
            let trailOptions = options.trail;
            if (!trailOptions?.load) {
                options.trail = trailOptions = new Trail();
            }
            trailOptions.load(source?.trail);
        }
        needsPlugin() {
            return true;
        }
    }

    async function loadTrailPlugin(engine) {
        engine.checkVersion("4.0.2");
        await engine.pluginManager.register(e => {
            e.pluginManager.addPlugin(new TrailPlugin(e.pluginManager));
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadTrailPlugin = loadTrailPlugin;

    const minimumLength = 0;
    class TrailPluginInstance {
        _container;
        _pluginManager;
        _trailFill;
        constructor(pluginManager, container) {
            this._container = container;
            this._pluginManager = pluginManager;
        }
        canvasClear() {
            const container = this._container, trail = container.actualOptions.trail, trailFill = this._trailFill;
            if (!trail?.enable || !trailFill || trail.length <= minimumLength) {
                return false;
            }
            let handled = false;
            const canvas = container.canvas;
            if (trailFill.color) {
                canvas.render.paintBase(engine.getStyleFromRgb(trailFill.color, container.hdr, trailFill.opacity));
                handled = true;
            }
            else if (trailFill.image) {
                canvas.render.paintImage(trailFill.image, trailFill.opacity);
                handled = true;
            }
            return handled;
        }
        async init() {
            try {
                await this._initTrail();
            }
            catch (e) {
                engine.getLogger().error(e);
            }
        }
        _initTrail = async () => {
            const options = this._container.actualOptions, trail = options.trail;
            if (!trail?.enable) {
                return;
            }
            const trailFill = trail.fill, opacity = engine.inverseFactorNumerator / trail.length;
            if (trailFill.color) {
                const fillColor = engine.rangeColorToRgb(this._pluginManager, trailFill.color);
                if (!fillColor) {
                    return;
                }
                this._trailFill = {
                    color: {
                        ...fillColor,
                    },
                    opacity,
                };
            }
            else {
                await new Promise((resolve, reject) => {
                    if (!trailFill.image) {
                        return;
                    }
                    const img = engine.safeDocument().createElement("img");
                    img.addEventListener("load", () => {
                        this._trailFill = {
                            image: img,
                            opacity,
                        };
                        resolve();
                    });
                    img.addEventListener("error", evt => {
                        engine.getLogger().error(evt);
                        reject(new Error("Error loading image"));
                    });
                    img.src = trailFill.image;
                });
            }
        };
    }

    var TrailPluginInstance$1 = /*#__PURE__*/Object.freeze({
        __proto__: null,
        TrailPluginInstance: TrailPluginInstance
    });

    exports.loadTrailPlugin = loadTrailPlugin;

}));
Object.assign(globalThis.window || globalThis, { loadTrailPlugin: (globalThis.__tsParticlesInternals.plugins.trail || {}).loadTrailPlugin });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
