// @tsparticles/plugin-poisson-disc/cjs/PoissonDisc.js
import { double, doublePI, getDistance, getRandom } from "@tsparticles/engine";
export class PoissonDisc {
    active;
    cellSize;
    cols;
    dimensions;
    firstPoint;
    grid;
    points;
    radius;
    retries;
    rows;
    size;
    constructor(size, radius, retries, dimensions, firstPoint) {
        this.size = { ...size };
        this.radius = radius;
        this.retries = retries;
        this.dimensions = dimensions;
        this.cellSize = Math.floor(this.radius / Math.sqrt(this.dimensions));
        this.cols = Math.floor(this.size.width / this.cellSize);
        this.rows = Math.floor(this.size.height / this.cellSize);
        this.points = [];
        this.active = [];
        this.grid = [];
        this.firstPoint = firstPoint ? { ...firstPoint } : undefined;
        this.reset();
    }
    addPoint(inputPoint) {
        const point = {
            position: { ...inputPoint },
            gridPosition: {
                x: Math.floor(inputPoint.x / this.cellSize),
                y: Math.floor(inputPoint.y / this.cellSize),
            },
        }, pointIndex = this.points.length, row = this.grid[point.gridPosition.y];
        if (!row) {
            return;
        }
        this.points.push(point);
        row[point.gridPosition.x] = pointIndex;
        this.active.push(pointIndex);
    }
    getRandom(min, max) {
        return Math.floor(getRandom() * (max - min)) + min;
    }
    initialiseGrid() {
        for (let y = 0; y <= this.rows; y++) {
            this.grid[y] = [];
            const row = this.grid[y];
            if (!row) {
                continue;
            }
            for (let x = 0; x <= this.cols; x++) {
                row[x] = -1;
            }
        }
    }
    reset() {
        this.points = [];
        this.active = [];
        this.grid = [];
        this.initialiseGrid();
        if (this.firstPoint) {
            this.addPoint(this.firstPoint);
        }
        else {
            const minCoordinate = 0;
            this.addPoint({
                x: this.getRandom(minCoordinate, this.size.width),
                y: this.getRandom(minCoordinate, this.size.height),
            });
        }
    }
    async run() {
        this.reset();
        const minCount = 0, step = 1, yieldEvery = 100, yieldStepModule = 0;
        let iterations = 0;
        while (this.active.length > minCount) {
            this.steps(step);
            if (++iterations % yieldEvery === yieldStepModule) {
                await new Promise(resolve => setTimeout(resolve));
            }
        }
    }
    steps(steps) {
        const minCount = 0;
        for (let i = 0; i < steps; i++) {
            if (this.active.length <= minCount) {
                continue;
            }
            this._step();
        }
    }
    _getNewPoint(currentPoint, tries) {
        const minCoordinate = 0, gridMinValue = 0, maxNeighbourIndex = 1, newAngle = tries * (doublePI / this.retries), newDist = this.getRandom(this.radius, this.radius * double), offset = {
            x: Math.cos(newAngle) * newDist,
            y: Math.sin(newAngle) * newDist,
        }, newPoint = {
            x: Math.floor(currentPoint.position.x + offset.x),
            y: Math.floor(currentPoint.position.y + offset.y),
        }, newGridCoords = {
            x: Math.floor(newPoint.x / this.cellSize),
            y: Math.floor(newPoint.y / this.cellSize),
        };
        if (newPoint.x <= minCoordinate ||
            newPoint.x >= this.size.width ||
            newPoint.y <= minCoordinate ||
            newPoint.y >= this.size.height) {
            return;
        }
        const row = this.grid[newGridCoords.y];
        if (!row) {
            return;
        }
        const cellValue = row[newGridCoords.x];
        if (cellValue === undefined) {
            return;
        }
        if (cellValue >= gridMinValue) {
            return;
        }
        for (let i = -1; i <= maxNeighbourIndex; i++) {
            for (let j = -1; j <= maxNeighbourIndex; j++) {
                if (!i && !j) {
                    continue;
                }
                const neighbourGrid = {
                    x: newGridCoords.x + j,
                    y: newGridCoords.y + i,
                };
                if (neighbourGrid.x < minCoordinate ||
                    neighbourGrid.y < minCoordinate ||
                    neighbourGrid.x >= this.cols ||
                    neighbourGrid.y >= this.rows) {
                    continue;
                }
                const neighbourCellValue = this.grid[neighbourGrid.y]?.[neighbourGrid.x];
                if (neighbourCellValue === undefined || neighbourCellValue < gridMinValue) {
                    continue;
                }
                const neighbour = this.points[neighbourCellValue];
                if (!neighbour) {
                    continue;
                }
                if (getDistance(newPoint, neighbour.position) < this.radius) {
                    return;
                }
            }
        }
        return newPoint;
    }
    _step() {
        const minCount = 0, randomActive = this.getRandom(minCount, this.active.length);
        let foundNewPoint = false;
        for (let tries = 0; tries < this.retries; tries++) {
            const randomActivePointIndex = this.active[randomActive];
            if (randomActivePointIndex === undefined) {
                continue;
            }
            const point = this.points[randomActivePointIndex];
            if (!point) {
                continue;
            }
            const newPoint = this._getNewPoint(point, tries);
            if (newPoint) {
                foundNewPoint = true;
                this.addPoint(newPoint);
                break;
            }
        }
        if (!foundNewPoint) {
            const deleteCount = 1;
            this.active.splice(randomActive, deleteCount);
        }
    }
}
