// @tsparticles/plugin-poisson-disc/tsparticles.plugin.poisson.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Plugin v4.0.2 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.plugins = global.__tsParticlesInternals.plugins || {}, global.__tsParticlesInternals.plugins.poisson = global.__tsParticlesInternals.plugins.poisson || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    class Poisson {
        dimensions;
        enable;
        radius;
        retries;
        steps;
        constructor() {
            this.enable = false;
            this.dimensions = 2;
            this.radius = 0;
            this.retries = 30;
            this.steps = 0;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.dimensions !== undefined) {
                this.dimensions = data.dimensions;
            }
            if (data.radius !== undefined) {
                this.radius = data.radius;
            }
            if (data.retries !== undefined) {
                this.retries = data.retries;
            }
        }
    }

    class PoissonDiscPlugin {
        id = "poisson";
        async getPlugin(container) {
            const { PoissonDiscPluginInstance } = await Promise.resolve().then(function () { return PoissonDiscPluginInstance$1; });
            return new PoissonDiscPluginInstance(container);
        }
        loadOptions(_container, options, source) {
            if (!this.needsPlugin(options) && !this.needsPlugin(source)) {
                return;
            }
            let poissonOptions = options.poisson;
            if (poissonOptions?.load === undefined) {
                options.poisson = poissonOptions = new Poisson();
            }
            poissonOptions.load(source?.poisson);
        }
        needsPlugin(options) {
            return options?.poisson?.enable ?? false;
        }
    }

    async function loadPoissonDiscPlugin(engine) {
        engine.checkVersion("4.0.2");
        await engine.pluginManager.register(e => {
            e.pluginManager.addPlugin(new PoissonDiscPlugin());
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadPoissonDiscPlugin = loadPoissonDiscPlugin;

    class PoissonDiscPluginInstance {
        poissonDisc;
        redrawTimeout;
        _container;
        _currentIndex;
        constructor(container) {
            this._container = container;
            this._currentIndex = 0;
        }
        async init() {
            await this._initData();
        }
        particlePosition(position) {
            const container = this._container, options = container.actualOptions.poisson;
            if (!this.poissonDisc || !(options?.enable ?? false) || this._currentIndex >= this.poissonDisc.points.length) {
                return;
            }
            return position ?? this.poissonDisc.points[this._currentIndex++]?.position;
        }
        resize() {
            const container = this._container, options = container.actualOptions.poisson;
            if (!(options?.enable ?? false)) {
                return;
            }
            if (this.redrawTimeout) {
                clearTimeout(this.redrawTimeout);
            }
            const timeout = 250;
            this.redrawTimeout = setTimeout(() => {
                void (async () => {
                    if (this._container.destroyed) {
                        return;
                    }
                    await this._initData();
                    await container.particles.redraw();
                })();
            }, timeout);
        }
        stop() {
            delete this.poissonDisc;
        }
        async _initData() {
            const container = this._container, poissonOptions = container.actualOptions.poisson, particlesOptions = container.actualOptions.particles, canvasSize = container.canvas.size, pixelRatio = container.retina.pixelRatio;
            if (!poissonOptions?.enable) {
                return;
            }
            this._currentIndex = 0;
            const { PoissonDisc } = await Promise.resolve().then(function () { return PoissonDisc$1; });
            this.poissonDisc = new PoissonDisc(canvasSize, poissonOptions.radius
                ? poissonOptions.radius * pixelRatio
                : Math.max(engine.getRangeMax(particlesOptions.size.value) * pixelRatio, Math.sqrt((canvasSize.width * canvasSize.height) / particlesOptions.number.value)), poissonOptions.retries, poissonOptions.dimensions);
            const noSteps = 0;
            if (poissonOptions.steps > noSteps) {
                this.poissonDisc.steps(poissonOptions.steps);
            }
            else {
                await this.poissonDisc.run();
            }
        }
    }

    var PoissonDiscPluginInstance$1 = /*#__PURE__*/Object.freeze({
        __proto__: null,
        PoissonDiscPluginInstance: PoissonDiscPluginInstance
    });

    class PoissonDisc {
        active;
        cellSize;
        cols;
        dimensions;
        firstPoint;
        grid;
        points;
        radius;
        retries;
        rows;
        size;
        constructor(size, radius, retries, dimensions, firstPoint) {
            this.size = { ...size };
            this.radius = radius;
            this.retries = retries;
            this.dimensions = dimensions;
            this.cellSize = Math.floor(this.radius / Math.sqrt(this.dimensions));
            this.cols = Math.floor(this.size.width / this.cellSize);
            this.rows = Math.floor(this.size.height / this.cellSize);
            this.points = [];
            this.active = [];
            this.grid = [];
            this.firstPoint = firstPoint ? { ...firstPoint } : undefined;
            this.reset();
        }
        addPoint(inputPoint) {
            const point = {
                position: { ...inputPoint },
                gridPosition: {
                    x: Math.floor(inputPoint.x / this.cellSize),
                    y: Math.floor(inputPoint.y / this.cellSize),
                },
            }, pointIndex = this.points.length, row = this.grid[point.gridPosition.y];
            if (!row) {
                return;
            }
            this.points.push(point);
            row[point.gridPosition.x] = pointIndex;
            this.active.push(pointIndex);
        }
        getRandom(min, max) {
            return Math.floor(engine.getRandom() * (max - min)) + min;
        }
        initialiseGrid() {
            for (let y = 0; y <= this.rows; y++) {
                this.grid[y] = [];
                const row = this.grid[y];
                if (!row) {
                    continue;
                }
                for (let x = 0; x <= this.cols; x++) {
                    row[x] = -1;
                }
            }
        }
        reset() {
            this.points = [];
            this.active = [];
            this.grid = [];
            this.initialiseGrid();
            if (this.firstPoint) {
                this.addPoint(this.firstPoint);
            }
            else {
                const minCoordinate = 0;
                this.addPoint({
                    x: this.getRandom(minCoordinate, this.size.width),
                    y: this.getRandom(minCoordinate, this.size.height),
                });
            }
        }
        async run() {
            this.reset();
            const minCount = 0, step = 1, yieldEvery = 100, yieldStepModule = 0;
            let iterations = 0;
            while (this.active.length > minCount) {
                this.steps(step);
                if (++iterations % yieldEvery === yieldStepModule) {
                    await new Promise(resolve => setTimeout(resolve));
                }
            }
        }
        steps(steps) {
            const minCount = 0;
            for (let i = 0; i < steps; i++) {
                if (this.active.length <= minCount) {
                    continue;
                }
                this._step();
            }
        }
        _getNewPoint(currentPoint, tries) {
            const minCoordinate = 0, gridMinValue = 0, maxNeighbourIndex = 1, newAngle = tries * (engine.doublePI / this.retries), newDist = this.getRandom(this.radius, this.radius * engine.double), offset = {
                x: Math.cos(newAngle) * newDist,
                y: Math.sin(newAngle) * newDist,
            }, newPoint = {
                x: Math.floor(currentPoint.position.x + offset.x),
                y: Math.floor(currentPoint.position.y + offset.y),
            }, newGridCoords = {
                x: Math.floor(newPoint.x / this.cellSize),
                y: Math.floor(newPoint.y / this.cellSize),
            };
            if (newPoint.x <= minCoordinate ||
                newPoint.x >= this.size.width ||
                newPoint.y <= minCoordinate ||
                newPoint.y >= this.size.height) {
                return;
            }
            const row = this.grid[newGridCoords.y];
            if (!row) {
                return;
            }
            const cellValue = row[newGridCoords.x];
            if (cellValue === undefined) {
                return;
            }
            if (cellValue >= gridMinValue) {
                return;
            }
            for (let i = -1; i <= maxNeighbourIndex; i++) {
                for (let j = -1; j <= maxNeighbourIndex; j++) {
                    if (!i && !j) {
                        continue;
                    }
                    const neighbourGrid = {
                        x: newGridCoords.x + j,
                        y: newGridCoords.y + i,
                    };
                    if (neighbourGrid.x < minCoordinate ||
                        neighbourGrid.y < minCoordinate ||
                        neighbourGrid.x >= this.cols ||
                        neighbourGrid.y >= this.rows) {
                        continue;
                    }
                    const neighbourCellValue = this.grid[neighbourGrid.y]?.[neighbourGrid.x];
                    if (neighbourCellValue === undefined || neighbourCellValue < gridMinValue) {
                        continue;
                    }
                    const neighbour = this.points[neighbourCellValue];
                    if (!neighbour) {
                        continue;
                    }
                    if (engine.getDistance(newPoint, neighbour.position) < this.radius) {
                        return;
                    }
                }
            }
            return newPoint;
        }
        _step() {
            const minCount = 0, randomActive = this.getRandom(minCount, this.active.length);
            let foundNewPoint = false;
            for (let tries = 0; tries < this.retries; tries++) {
                const randomActivePointIndex = this.active[randomActive];
                if (randomActivePointIndex === undefined) {
                    continue;
                }
                const point = this.points[randomActivePointIndex];
                if (!point) {
                    continue;
                }
                const newPoint = this._getNewPoint(point, tries);
                if (newPoint) {
                    foundNewPoint = true;
                    this.addPoint(newPoint);
                    break;
                }
            }
            if (!foundNewPoint) {
                const deleteCount = 1;
                this.active.splice(randomActive, deleteCount);
            }
        }
    }

    var PoissonDisc$1 = /*#__PURE__*/Object.freeze({
        __proto__: null,
        PoissonDisc: PoissonDisc
    });

    exports.loadPoissonDiscPlugin = loadPoissonDiscPlugin;

}));
Object.assign(globalThis.window || globalThis, { loadPoissonDiscPlugin: (globalThis.__tsParticlesInternals.plugins.poisson || {}).loadPoissonDiscPlugin });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
