// @tsparticles/preset-sea-anemone/tsparticles.preset.seaAnemone.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Preset v4.0.2 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/basic'), require('@tsparticles/path-curves'), require('@tsparticles/plugin-emitters/plugin'), require('@tsparticles/plugin-trail')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/basic', '@tsparticles/path-curves', '@tsparticles/plugin-emitters/plugin', '@tsparticles/plugin-trail'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.presets = global.__tsParticlesInternals.presets || {}, global.__tsParticlesInternals.presets.seaAnemone = global.__tsParticlesInternals.presets.seaAnemone || {}), global.__tsParticlesInternals.bundles.basic, global.__tsParticlesInternals.paths.curves, global.__tsParticlesInternals.plugins["emitters/plugin"], global.__tsParticlesInternals.plugins.trail));
})(this, (function (exports, basic, pathCurves, plugin, pluginTrail) { 'use strict';

    const options = {
        fpsLimit: 120,
        particles: {
            paint: {
                color: {
                    value: "#FF0000",
                },
            },
            move: {
                direction: "none",
                enable: true,
                outModes: {
                    default: "destroy",
                },
                path: {
                    clamp: false,
                    enable: true,
                    delay: {
                        value: 0,
                    },
                    generator: pathCurves.curvesPathName,
                },
                random: false,
                speed: 2,
                straight: false,
            },
            number: {
                value: 0,
                limit: { value: 300 },
            },
            opacity: {
                value: 1,
            },
            shape: {
                type: "circle",
            },
            size: {
                value: { min: 1, max: 10 },
                animation: {
                    count: 1,
                    startValue: "min",
                    enable: true,
                    speed: 10,
                    sync: true,
                },
            },
        },
        background: {
            color: "#000",
        },
        detectRetina: true,
        emitters: {
            direction: "none",
            rate: {
                quantity: 10,
                delay: 0.3,
            },
            size: {
                width: 0,
                height: 0,
                mode: "precise",
            },
            spawn: {
                fill: {
                    color: {
                        value: "#ff0000",
                        animation: {
                            h: {
                                enable: true,
                                offset: {
                                    min: -1.4,
                                    max: 1.4,
                                },
                                speed: 5,
                                sync: false,
                            },
                            l: {
                                enable: true,
                                offset: {
                                    min: 20,
                                    max: 80,
                                },
                                speed: 0,
                                sync: false,
                            },
                        },
                    },
                },
            },
            position: {
                x: 50,
                y: 50,
            },
        },
        trail: {
            fill: {
                color: "#000",
            },
            length: 30,
            enable: true,
        },
    };

    const presetName = "seaAnemone";
    async function loadSeaAnemonePreset(engine) {
        await engine.pluginManager.register(async (e) => {
            await Promise.all([
                (async () => {
                    await basic.loadBasic(e);
                    await pathCurves.loadCurvesPath(e);
                })(),
                plugin.loadEmittersPluginSimple(e),
                pluginTrail.loadTrailPlugin(e),
            ]);
            e.pluginManager.addPreset(presetName, options);
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadSeaAnemonePreset = loadSeaAnemonePreset;

    exports.loadSeaAnemonePreset = loadSeaAnemonePreset;

}));
Object.assign(globalThis.window || globalThis, { loadSeaAnemonePreset: (globalThis.__tsParticlesInternals.presets.seaAnemone || {}).loadSeaAnemonePreset });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
