// @tsparticles/shape-matrix/tsparticles.shape.matrix.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Shape v4.0.2 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.shapes = global.__tsParticlesInternals.shapes || {}, global.__tsParticlesInternals.shapes.matrix = global.__tsParticlesInternals.shapes.matrix || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    const matrixChars = [
        "\u30A2",
        "\u30A4",
        "\u30A6",
        "\u30A8",
        "\u30AA",
        "\u30AB",
        "\u30AD",
        "\u30AF",
        "\u30B1",
        "\u30B3",
        "\u30B5",
        "\u30B7",
        "\u30B9",
        "\u30BB",
        "\u30BD",
        "\u30BF",
        "\u30C1",
        "\u30C4",
        "\u30C6",
        "\u30C8",
        "\u30CA",
        "\u30CB",
        "\u30CC",
        "\u30CD",
        "\u30CE",
        "\u30CF",
        "\u30D2",
        "\u30D5",
        "\u30D8",
        "\u30DB",
        "\u30DE",
        "\u30DF",
        "\u30E0",
        "\u30E1",
        "\u30E2",
        "\u30E4",
        "\u30E6",
        "\u30E8",
        "\u30E9",
        "\u30EA",
        "\u30EB",
        "\u30EC",
        "\u30ED",
        "\u30EF",
        "\u30F2",
        "\u30F3",
        "\u30AC",
        "\u30AE",
        "\u30B0",
        "\u30B2",
        "\u30B4",
        "\u30B6",
        "\u30B8",
        "\u30BA",
        "\u30BC",
        "\u30BE",
        "\u30C0",
        "\u30C2",
        "\u30C5",
        "\u30C7",
        "\u30C9",
        "\u30D0",
        "\u30D3",
        "\u30D6",
        "\u30D9",
        "\u30DC",
        "\u30D1",
        "\u30D4",
        "\u30D7",
        "\u30DA",
        "\u30DD",
        "\u30A1",
        "\u30A3",
        "\u30A5",
        "\u30A7",
        "\u30A9",
        "\u30E3",
        "\u30E5",
        "\u30E7",
        "\u30C3",
        "\u30FC",
        "\u30F4",
        "0",
        "1",
        "2",
        "3",
        "4",
        "5",
        "6",
        "7",
        "8",
        "9",
        "\uFF10",
        "\uFF11",
        "\uFF12",
        "\uFF13",
        "\uFF14",
        "\uFF15",
        "\uFF16",
        "\uFF17",
        "\uFF18",
        "\uFF19",
        "A",
        "B",
        "C",
        "D",
        "E",
        "F",
        "G",
        "H",
        "I",
        "J",
        "K",
        "L",
        "M",
        "N",
        "O",
        "P",
        "Q",
        "R",
        "S",
        "T",
        "U",
        "V",
        "W",
        "X",
        "Y",
        "Z",
        "\u65E5",
        "\u672C",
        "\u4EBA",
        "\u5927",
        "\u4E2D",
        "\u5C0F",
        "\u4E0A",
        "\u4E0B",
        "\u5DE6",
        "\u53F3",
        "\u529B",
        "\u6642",
        "\u7A7A",
        "\u751F",
        "\u6B7B",
        "\u7121",
        "\u6709",
        "!",
        "@",
        "#",
        "$",
        "%",
        "&",
        "*",
        "+",
        "-",
        "=",
        "?",
        "/",
        "\\",
        "|",
        "<",
        ">",
        "{",
        "}",
        "[",
        "]",
        "(",
        ")",
    ];
    const defaultMinInterval = 100, defaultMaxInterval = 2000, charCache = new WeakMap();
    function randomChar() {
        return matrixChars[Math.floor(engine.getRandom() * matrixChars.length)] ?? " ";
    }
    function randomInterval(minInterval, maxInterval) {
        return minInterval + engine.getRandom() * (maxInterval - minInterval);
    }
    function getState(particle, minInterval, maxInterval) {
        let state = charCache.get(particle);
        if (!state) {
            state = {
                char: randomChar(),
                elapsed: 0,
                interval: randomInterval(minInterval, maxInterval),
            };
            charCache.set(particle, state);
        }
        return state;
    }
    function drawMatrix(data) {
        const { context, radius, particle, delta, fill, stroke } = data, shapeData = particle.shapeData, minInterval = engine.getRangeMin(shapeData?.interval ?? defaultMinInterval), maxInterval = engine.getRangeMax(shapeData?.interval ?? defaultMaxInterval), state = getState(particle, minInterval, maxInterval);
        state.elapsed += delta.value;
        if (state.elapsed >= state.interval) {
            state.char = randomChar();
            state.elapsed -= state.interval;
            state.interval = randomInterval(minInterval, maxInterval);
        }
        const fontSize = Math.round(radius * engine.double);
        context.font = `bold ${fontSize}px "Courier New", Courier, monospace`;
        context.textAlign = "center";
        context.textBaseline = "middle";
        if (fill) {
            context.fillText(state.char, engine.originPoint.x, engine.originPoint.y);
        }
        if (stroke) {
            context.strokeText(state.char, engine.originPoint.x, engine.originPoint.y);
        }
    }

    class MatrixDrawer {
        draw(data) {
            drawMatrix(data);
        }
    }

    async function loadMatrixShape(engine) {
        engine.checkVersion("4.0.2");
        await engine.pluginManager.register(e => {
            e.pluginManager.addShape(["matrix"], () => Promise.resolve(new MatrixDrawer()));
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadMatrixShape = loadMatrixShape;

    exports.loadMatrixShape = loadMatrixShape;

}));
Object.assign(globalThis.window || globalThis, { loadMatrixShape: (globalThis.__tsParticlesInternals.shapes.matrix || {}).loadMatrixShape });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
